go/vt/vtgate/planbuilder/operators/horizon_planning.go

```diff
@@ -19,60 +19,196 @@ package operators
+func planColumns(ctx *plancontext.PlanningContext, root ops.Operator) (ops.Operator, error) {
+	// We only need to do column planning to the point we hit a Route.
+	// Everything underneath the route is handled by the mysql planner
+	stopAtRoute := func(operator ops.Operator) rewrite.VisitRule {
+		_, isRoute := operator.(*Route)
+		return rewrite.VisitRule(!isRoute)
+	}
+	visitor := func(in ops.Operator, _ semantics.TableSet) (ops.Operator, rewrite.TreeIdentity, error) {
 		switch in := in.(type) {
 		case *Horizon:
-			op, visit, err := planHorizon(ctx, in)
+			op, err := planHorizon(ctx, in, in == root)
+			if err != nil {
+				return nil, false, err
+			}
+			return op, rewrite.NewTree, nil
+		case *Derived:
+			op, err := planDerived(ctx, in, in == root)
 			if err != nil {
-				return nil, rewrite.SameTree, rewrite.SkipChildren, err
+				return nil, false, err
 			}
-			return op, rewrite.NewTree, visit, nil
-		case *Route:
-			return in, rewrite.SameTree, rewrite.SkipChildren, nil
+			return op, rewrite.NewTree, err
+		case *Filter:
+			err := planFilter(ctx, in)
+			if err != nil {
+				return nil, false, err
+			}
+			return in, rewrite.SameTree, nil
 		default:
-			return in, rewrite.SameTree, rewrite.VisitChildren, nil
+			return in, rewrite.SameTree, nil
 		}
-	})
+	}
+
+	newOp, err := rewrite.BottomUp(root, TableID, visitor, stopAtRoute)
+	if err != nil {
+		if vterr, ok := err.(*vterrors.VitessError); ok && vterr.ID == "VT13001" {
+			// we encountered a bug. let's try to back out
+			return nil, errNotHorizonPlanned
+		}
+		return nil, err
+	}
+
+	return newOp, nil
 }
 
-func planHorizon(ctx *plancontext.PlanningContext, in *Horizon) (ops.Operator, rewrite.VisitRule, error) {
+func planHorizon(ctx *plancontext.PlanningContext, in *Horizon, isRoot bool) (ops.Operator, error) {
 	rb, isRoute := in.Source.(*Route)
-	if !isRoute {
-		return in, rewrite.VisitChildren, nil
+	if isRoot && !isRoute && ctx.SemTable.NotSingleRouteErr != nil {
+		// If we got here, we don't have a single shard plan
+		return nil, ctx.SemTable.NotSingleRouteErr
 	}
 	if isRoute && rb.IsSingleShard() && in.Select.GetLimit() == nil {
 		return planSingleRoute(rb, in)
 	}
 
 	sel, isSel := in.Select.(*sqlparser.Select)
 	if !isSel {
-		return nil, rewrite.VisitChildren, errNotHorizonPlanned
+		return nil, errNotHorizonPlanned
 	}
 
 	qp, err := CreateQPFromSelect(ctx, sel)
 	if err != nil {
-		return nil, rewrite.VisitChildren, err
+		return nil, err
 	}
 
 	needsOrdering := len(qp.OrderExprs) > 0
 	canShortcut := isRoute && sel.Having == nil && !needsOrdering
+	_, isDerived := in.Source.(*Derived)
 
-	if !qp.NeedsAggregation() && sel.Having == nil && canShortcut && !needsOrdering && !qp.NeedsDistinct() && in.Select.GetLimit() == nil {
+	switch {
+	case qp.NeedsAggregation() || sel.Having != nil || sel.Limit != nil || isDerived || needsOrdering || qp.NeedsDistinct():
+		return nil, errNotHorizonPlanned
```

Comment:
so this is some of the cases we do not handle in the Horizon Planning Operator today?

Reply:
exactly